### ISSUE TITLE
URL fails to parse

### ISSUE THREAD

**Issue by thomasmey** [rust-lang/rust#140129]
Hi,

I'm completely new to rust and tried to setup rust with rustup tool in an corporate environment. I did setup RUSTUP_DIST_SERVER to the intern mirror, which sadly needs basic authentication.

I used an URL like this:
https://user@example.com:P#ssword@mirror.example.com

Strangely rustup does throw an error like: invalid port number.

After some fiddling it looks like that the # character in the password is the reason for this error, when I remove it the URL get's parsed correctly.
Bit even sure if basic authentication is added by rustup tool even if specified in RUSTUP_DIST_SERVER.

This happens on windows 10.

I tried this code:

```rust
<code>
```

I expected to see this happen: *explanation*

Instead, this happened: *explanation*

### Meta

`rustc --version --verbose`:
```
<version>
```

<details><summary>Backtrace</summary>
<p>

```
<backtrace>
```

</p>
</details>
